README.md
<h1 align="center">mBlog</h1><br>

[![GitHub](https://img.shields.io/github/license/wgrocha/mblog?style=for-the-badge)](https://github.com/wgrocha/mblog/blob/master/LICENSE)
[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/wgrocha/mblog?sort=semver&style=for-the-badge)](https://github.com/wgrocha/mblog/releases)

### Table of contents

- [What is mBlog?](#what-is-mblog) 
- [How it works?](#how-it-works)
- [Download and setup](#download-and-setup)
- [Usage](#usage)
- [Plugins](#plugins)
- [Contributing](#contributing)
- [Credits](#credits)
- [License](#license)


##### Other

- [Documentation](DOCUMENTATION.md)
- Official mBlog site __*(not available yet)*__
- FAQ __*(not available yet)*__

#

### What is mBlog?

mBlog is a lightweight blog engine, it was made to be simple, fast, extensible, dynamic and it is totally open-source! 
Besides that it doesn't even use any server-side languages. It's perfect to be used on [GitHub Pages](http://pages.github.com) :wink:

You can look at mBlog like a CMS, the simplest CMS you can imagine. Don't waste time comparing it to other CMS platforms like Wordpress, Joomla, or any other, mBlog was designed to be a simple blog engine and nothing more. 

[Click here to check a live demo of mBlog.](https://wgrocha.github.io/mblog-demo/)

### How it works?

Because mBlog does't use server-side languages, it abuses Javascript, all content is loaded using JS. There are no frameworks, it is just raw Javascript, pure and simple. The articles are all saved in a JSON file __*(articles.json)*__ which can be found inside the __*data*__ folder. For each new article you write you will download a new __*articles.json*__ file, you just have to replace the old one and _voilà_: published!

### Download and setup

To download and setup mBlog is very simple, there are three methods to do this:

- You can access the official download page, there you must set a few things, for that you just fill a small form, pass some basic information like your blog description, your name, etc. Doing this you will download mBlog configured and ready to be used! __*(This method is not available yet!)*__

- You can clone this GitHub repository, inside of it you'will find a small python script, the script name is __*config.py*__, just run it (you need to have python installed) and it will help you to configure things like your blog description, your name, links to your GitHub, Linkedin, etc:
```
git clone https://github.com/wgrocha/mblog.git
cd mblog
python3 config.py
```
- If you don't want do clone the repository or download from the download page, you can download [right here](https://github.com/wgrocha/mblog/archive/master.zip), extract the files and run the __*config.py*__ script.

### Usage

With mBlog already configured you can start using it by uploading the files to a web server.<br>
We recommend to use mBlog on GitHub Pages!

##### How to write a new article?

To write and post an article is very simple, you don't even need to access complicated dashboards. Follow these steps to write and publish an article:

- After uploading the files to a web server go to your mBlog homepage, there is a button that says "new article" or something like that, just click on it and you will be redirected to a new blank article;

- When the page for the new article is loaded, a pencil will appear in the upper left corner of the screen, clicking on this pencil you enable the editing mode so you can edit the new article as you wish, add text, images, videos, tables, lists and you can even edit HTML code directly (for more advanced users);

- In addition, it is possible to add a banner to the new article, the recommended resolution is 1000x300px, just paste the URL of the image in the corresponding field during the creation of the article and the banner will be automatically loaded;

- After writing the article, add tags to it at the end of the page, tags can be used by plugins to filter your articles, for example;

- Finally click on the button that indicates you're done. The page will ask you to download a file called __*articles.json*__, this file is the new database already updated with the new article, just replace the old __*articles.json*__ (inside __*data*__ folder) and voilà: article created and published.

Can you imagine this on GitHub Pages? Posting a new article is as easy as running a _git push_!

##### How the new article editor works?

The editor used to edit content and write new articles is called __*ContentTools*__.<br>
[Here](https://getcontenttools.com/demo) you access the official guide explaining how to use it.

### Plugins

Yeeep! mBlog supports plugins.

##### How to install a plugin?

Installing plugins couldn't be simpler:

1. Download the plugin (it's going to be a single __*.js*__ file);
2. Put the plugin inside the __*plugins*__ folder;
3. Open the __*plugins.json*__ file (inside the __*data*__ folder), and add the name of the new plugin to it.

Below you can see an example of __*plugins.json*__ file:
```
[
	"social-media-share.js",
	"another-plugin.js",
	"one-more-plugin.js"
]
```
By default mBlog comes with one activated plugin __*(social-media-share.js)*__, it adds links to the end of each article which enable the reader to share it on social media.

##### How to create a plugin?

If you want to create a plugin for mBlog please visit the [documentation](DOCUMENTATION.md) and go get things done :wink:

### Contributing

Contributions are always welcome! Please read the [contribution guidelines](CONTRIBUTING.md) first.

### Credits

[ContentTools](https://getcontenttools.com) - Without this amazing tool mBlog would not be possible :blue_heart:


### License

This work by [Willian Gabriel C. da Rocha](https://wgrocha.github.io) is licensed under MIT License.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2020-06-24

### Added
- ```new_version.py``` - Script to change version number in files for a new release (for developers)
- ```DOCUMENTATION.md``` - The technical description of the project
- Plugin support for the "Write Article" page _(mblog/write/index.html)_
- ```.gitignore``` to prevent tracking of _articles.json_ file
- This CHANGELOG file

### Changed
- Updated instructions in the _Contribution Guidelines_


## [0.1.0] - 2020-06-22
- mBlog was born!


config.py
from os import system, name

#Replace constans in each html file
def replace_data (constant, content):
	#Files where data will be replaced
	files = ['index.html', 'article.html', './write/index.html']
	#Goes through the array
	for file_name in files:
		with open (file_name) as f:
			newText = f.read().replace (constant, content)
		with open (file_name, "w") as f:
			f.write (newText)
#End of replace_data

#Verify if all user inputs are valid
def verify_inputs (name, bio, github, linkedin, email):
	#Still have to implement
	return True
#End of verify_inputs

#To clear the screen
def cls (): 
	# for windows 
	if name == 'nt': 
		_ = system('cls') 
	# for mac and linux (here, os.name is 'posix') 
	else: 
		_ = system('clear')
#End of cls


#The entrypoint of the script
def main ():
	correct = 'n'
	while (correct.lower()[0] not in ['y', 'Y']):
		cls ()
		print ('Hi, i\'m going to help you configure your mBlog!\nI have a few questions for you. Let\'s start!');
		name = input ('\nWhat\'s the author\'s name for the blog? ')
		bio = input ('Tell me a simple bio of the author (this is going to be visible in the home page, be concise): ')
		github = input ('Now give the link to your Github (You can leave it blank if you want): ')
		linkedin = input ('Now the link to your LinkedIn (You can also leave blank): ')
		email = input ('We\'re almost there, now I need an email for people to contact you! (You can also leave blank): ')
		
		#Verify inputs and get confirmation from user
		if (verify_inputs (name, bio, github, linkedin, email) == False):
			_ = input ('\nOps, something is wrong.\nPlease verify your inputs and try again! (Type ENTER to continue) ')
		else:
			print ('\nI\'m done with questions, please check all the information below:\n')
			print ('Name:', name)
			print ('Bio:', bio)
			print ('Github:', github)
			print ('LinkedIn:', linkedin)
			print ('Email:', email)
			correct = input ('\nIs this correct? (Y/N): ')
		#End of if/else
	#End of while
	
	constants = {
		'{AUTHOR_NAME}': name, 
		'{AUTHOR_BIO}': bio, 
		'{AUTHOR_GITHUB}': github, 
		'{AUTHOR_LINKEDIN}': linkedin, 
		'{AUTHOR_EMAIL}': email
	}

	#Iterate replacing data on mBlog files
	for key, value in constants.items ():
		replace_data (key, value)
	#End of for
	
	#End of setup
	print ('\nAwesome! We\'ve finished the configuration of your blog ;)\nNow you can delete me, upload all other files to a server and start writing!')
#End of main




#Start everything
main ()

new_version.py
#!/usr/bin/python3

import sys, getopt

#Replace version number in html files
def replace_version (current_version, new_version):
	#Files where version number will be replaced
	files = ['index.html', 'article.html', './write/index.html']
	#Goes through the array replacing the version in each file
	for file_name in files:
		with open (file_name) as f:
			newFileContent = f.read ().replace (current_version, new_version)
		with open (file_name, 'w') as f:
			f.write (newFileContent)


#Parses and validates command line arguments
def parse_arguments (argv):
	from_version = ''
	to_version = ''
	try:
		opts, args = getopt.getopt (sys.argv[1:], '',['from=','to='])
	except getopt.GetoptError:
		print ('Usage: new_version.py --from <current_version> --to <new_version>')
		sys.exit ()
	for opt, arg in opts:
		if opt == '--help':
			print ('Usage: new_version.py --from <current_version> --to <new_version>')
			sys.exit ()
		elif opt in ('--from'):
			from_version = arg
		elif opt in ('--to'):
			to_version = arg
		else:
			print ('Usage: new_version.py --from <current_version> --to <new_version>')
			sys.exit ()
	if (from_version == '' or to_version == ''):
		print ('Usage: new_version.py --from <current_version> --to <new_version>')
		sys.exit ()
	else:
		#Returns parsed arguments
		return (from_version, to_version)


# Parses command line arguments --from and --to so executes the 
if __name__ == '__main__':
	from_version, to_version = parse_arguments (sys.argv)
	replace_version (from_version, to_version);

js/article-list.js
var currentArticlePage = 0;
var articlesPerPage = 5;

document.addEventListener ("DOMContentLoaded", function loadArticleList () {
    //To decode base64
    function decodeUnicode (str) {
        // Going backwards: from bytestream, to percent-encoding, to original string.
        return decodeURIComponent(atob(str).split('').map(function (c) {
            return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
        }).join(''));
    }
    //Used to load database file
    function loadDatabase (file, callback) {
        let xhr = new XMLHttpRequest ();
        xhr.overrideMimeType ("application/json");
        xhr.open ("GET", file, true);
        xhr.onreadystatechange = () => {
            if (xhr.readyState === XMLHttpRequest.DONE) {
                if (xhr.status === 0 || (xhr.status >= 200 && xhr.status < 400)) {
                    try {
                        callback (JSON.parse (xhr.responseText));
                    } catch (e) {
                        document.getElementById ('list-of-articles').innerHTML = '<p class="align-center" style="color: rgba(0, 0, 0, 0.5)">No articles yet!</p>';    
                    }
                } else {
                    document.getElementById ('list-of-articles').innerHTML = '<p class="align-center" style="color: rgba(0, 0, 0, 0.5)">No articles yet!</p>';
                }
            }
        }
        xhr.send (null);
    }
    loadDatabase ('./data/articles.json', (database) => {
        //Check if all articles can be displayed in a single page
        if (database.length <= articlesPerPage) {
            document.getElementById ('pages-section').parentNode.removeChild (document.getElementById ('pages-section'));
        } else {
            //Calculate the number of pages
            let numPages = Math.ceil (database.length / articlesPerPage);
            //Create buttons for page navigation
            document.getElementById ('page-buttons').innerHTML = '';
            for (let i = 0; i < numPages; i++) {
                let new_button = document.createElement ('button');
                new_button.classList.add ('article-nav-button');
                if (i == currentArticlePage) new_button.classList.add ('button-highlight')
                new_button.innerHTML = i+1;
                new_button.addEventListener ('click', () => {
                    currentArticlePage = i;
                    document.getElementById ('articles-section-top').scrollIntoView ({
                        behavior: "smooth",
                        block: "start"
                    });
                    loadArticleList ();
                });
                document.getElementById ('page-buttons').appendChild (new_button);
            }
            //Calculate first and last article to show in the current page
            let first = articlesPerPage * currentArticlePage;
            let last = database.length -1;
            if (articlesPerPage <= database.length) {
                last = articlesPerPage * currentArticlePage + articlesPerPage;
            }
            //Get only articles for the current page
            database = database.slice (first, last);
        }
        //Clean the div where the article list will appear
        document.getElementById ('list-of-articles').innerHTML = '';
        //Show the article list
        database.forEach ((article) => {
            let articleTags = document.createElement ('span');
            let tagList = article.articleTags;
            tagList.forEach ((tag) => {
                new_tag = document.createElement ('p');
                new_tag.innerHTML = tag;
                articleTags.appendChild (new_tag);
            });
            if (tagList.length == 0) {
                let message = document.createElement ('p');
                message.innerHTML = 'Untagged';
                articleTags.appendChild (message);
            }
            let sectionTemplate = [
                "<section class=\"article\">",
                    "<div class=\"row\">",
                        "<div class=\"container\">",
                            "<div class=\"row\">",
                                "<div class=\"title xlarge-8 xlarge-offset-3 large-8 large-offset-3 medium-8 medium-offset-3 align-middle\">",
                                    article.articleTitle,
                                "</div>",
                                "<div class=\"subtitle xlarge-8 xlarge-offset-3 large-8 large-offset-3 medium-8 medium-offset-3 align-middle\">",
                                    article.articleSubtitle,
                                "</div>",
                                "<div class=\"date xlarge-8 xlarge-offset-3 large-8 large-offset-3 medium-8 medium-offset-3 align-middle\">",
                                    article.articleDate,
                                "</div>",
                                "<div class=\"summary xlarge-8 xlarge-offset-3 large-8 large-offset-3 medium-8 medium-offset-3\">",
                                    decodeUnicode (article.articleSummary),
                                    " <a href=\"article.html?id=",
                                        article.articleID,
                                    "\" target=\"_blank\">Read more...</a>",
                                "</div>",
                                "<div class=\"tags xlarge-8 xlarge-offset-3 large-8 large-offset-3 medium-8 medium-offset-3\"><span style=\"font-weight:bold; font-size: 14px;\">Tags:</span>",
                                    articleTags.innerHTML,
                                "</div>",
                            "</div>",
                        "</div>",
                    "</div>",
                    "<div class=\"section-divider\"></div>",
                "</section>"
            ];
            document.getElementById ('list-of-articles').innerHTML += sectionTemplate.join('');
        });

    });
});

js/plugin-loader.js
document.addEventListener ('DOMContentLoaded', () => {
	//Used to load database file
	function getPluginList (file, callback) {
		let xhr = new XMLHttpRequest ();
		xhr.overrideMimeType ("application/json");
		xhr.open ("GET", file, true);
		xhr.onreadystatechange = () => {
			if (xhr.readyState === XMLHttpRequest.DONE) {
				if (xhr.status === 0 || (xhr.status >= 200 && xhr.status < 400)) {
					callback (JSON.parse (xhr.responseText));
				}
			}
		}
		xhr.send (null);
	}

	//Check if called from inside the write folder
	let plugins_json = 'data/plugins.json';
	if (document.location.toString().indexOf('/write/')) {
		plugins_json = '../data/plugins.json';
	}
	//Get plugin list and add them inside the head tag
	getPluginList (plugins_json, (plugins) => {
		if (plugins) {
			let head = document.getElementsByTagName ('head')[0];
			plugins.forEach ((plugin) => {
				//Adding the script tag to the head
				let script = document.createElement ('script');
				script.type = 'text/javascript';
				if (document.location.toString().indexOf('/write/')) {
					script.src = '../plugins/' + plugin;
				} else {
					script.src = 'plugins/' + plugin;
				}
				head.appendChild (script);
			});
		}
	});
});

js/article-loader.js
//This script loads an entire article inside article.html page
document.addEventListener ("DOMContentLoaded", loadArticle);

//To decode base64
function decodeUnicode(str) {
  // Going backwards: from bytestream, to percent-encoding, to original string.
  return decodeURIComponent(atob(str).split('').map(function (c) {
    return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
  }).join(''));
}

//Load articles from database
function loadArticle () {
	//Used to load database file
	function loadDatabase (file, callback) {
		let xhr = new XMLHttpRequest ();
		xhr.overrideMimeType ("application/json");
		xhr.open ("GET", file, true);
		xhr.onreadystatechange = () => {
			if (xhr.readyState === XMLHttpRequest.DONE) {
				if (xhr.status === 0 || (xhr.status >= 200 && xhr.status < 400)) {
					try {
						callback (JSON.parse (xhr.responseText));	
					} catch (e) {
						document.getElementById ('article-title').innerHTML = '<p>Ops!</p>';
						document.getElementById ('article-subtitle').innerHTML = '<p>No articles yet. Come back later ;)</p>';
						document.getElementById ('social-media-share').style.display = 'none';	
					}
				} else {
					document.getElementById ('article-title').innerHTML = '<p>Ops!</p>';
					document.getElementById ('article-subtitle').innerHTML = '<p>No articles yet. Come back later ;)</p>';
					document.getElementById ('social-media-share').style.display = 'none';
				}
			}
		}
		xhr.send (null);
	}
	//To obtain get variables from the url
	let $_GET = {};
	if (document.location.toString().indexOf('?') !== -1) {
		let query = document.location
		.toString()
		.replace(/^.*?\?/, '')
		.replace(/#.*$/, '')
		.split('&');
		for (let i = 0, l = query.length; i < l; i++) {
			let aux = decodeURIComponent(query[i]).split('=');
			$_GET[aux[0]] = aux[1];
		}
	}
	//Get the database file
	loadDatabase ('./data/articles.json', (database) => {
		let articleExists = false;
		database.forEach ((article) => {
    		if ($_GET['id'] && article.articleID == $_GET['id']) {
    			articleExists = true;
				//Set article title
				let title = document.createElement ('h2');
				title.innerHTML = article.articleTitle;
				document.getElementById ('article-title').appendChild (title);
				//Set article subtitle
				let subtitle = document.createElement ('h3');
				subtitle.innerHTML = article.articleSubtitle;
				document.getElementById ('article-subtitle').appendChild (subtitle);
				//Set article date
				let date = document.createElement ('p');
				date.innerHTML = article.articleDate;
				document.getElementById ('article-date').appendChild (date);
				//Set article image
				if (article.articleImage) {
					document.getElementById ('article-image').style.backgroundImage = "url('" + article.articleImage + "')";
				} else {
					document.getElementById ('article-image').parentNode.removeChild (document.getElementById ('article-image'));
				}
				//Set article text
				document.getElementById ('article-text').innerHTML = decodeUnicode (article.articleText);
				//Set article author
				let author = document.createElement ('p');
				author.innerHTML = article.articleAuthor;
				document.getElementById ('article-author').appendChild (author);
				//Set article tags
				let tags = article.articleTags;
				tags.forEach ((tag) => {
					let new_tag = document.createElement ('p');
					new_tag.innerHTML = tag;
					document.getElementById ('article-tags').appendChild (new_tag);
				});
			}
		});
		//If the article does not exists shows an error message
		if (articleExists == false) {
			document.getElementById ('article-title').innerHTML = '<p>404 Ops!</p>';
			document.getElementById ('article-subtitle').innerHTML = '<p>This article doesn\'t exists.</p>';
		}
	});
}

plugins/social-media-share.js
/* Social media share buttons */
//Must run only on article.html
if (window.location.href.indexOf ('article.html') != -1) {
	//Create a new div after the article
	let share_div = document.createElement ('div');
	share_div.style.fontSize = '17px';
	share_div.style.marginTop = '50px';
	share_div.classList.add ('xlarge-12');
	share_div.classList.add ('align-center');
	//Add to page
	document.getElementById ('article').appendChild (share_div);
	//Create buttons
	let thisURL = window.location.href;
	let shareLabel = "<p style='margin-bottom: 15px; font-size: 14px;'>Share on social media:</p>";
	let facebookButton = "<a style='display: inline-block; font-family: Helvetica, arial; color: #fff; font-size: 14px; border-radius: 4px; padding: 2px 6px 2px 6px; margin: 2px; background-color: #4267b2;' href=https://www.facebook.com/sharer/sharer.php?u=" + thisURL + " target=_blank>Facebook</a>";
	let twitterButton = "<a style='display: inline-block; font-family: Helvetica, arial; color: #fff; font-size: 14px; border-radius: 4px; padding: 2px 6px 2px 6px; margin: 2px; background-color: #38A1F3;' href=https://twitter.com/home?status=" + thisURL + " target=_blank><span class=icon icon-twitter></span>Twitter</a>";
	let linkedinButton = "<a style='display: inline-block; font-family: Helvetica, arial; color: #fff; font-size: 14px; border-radius: 4px; padding: 2px 6px 2px 6px; margin: 2px; background-color: #0077B5;' href=https://www.linkedin.com/shareArticle?mini=true&amp;url=" + thisURL + " target=_blank>LinkedIn</a>";
	let copyButton = "<a id='copy-share-link' style='display: inline-block; font-family: Helvetica, arial; color: #fff; font-size: 14px; border-radius: 4px; padding: 2px 6px 2px 6px; margin: 2px; background-color: #993232;'>Copy URL</a>";
	let whatsappButton = "<a style='display: inline-block; font-family: Helvetica, arial; color: #fff; font-size: 14px; border-radius: 4px; padding: 2px 6px 2px 6px; margin: 2px; background-color: #25d366;' href='whatsapp://send?text=" + thisURL + "'>WhatsApp</a>";
	//Add buttons to div
	share_div.innerHTML = shareLabel + facebookButton + whatsappButton + linkedinButton + twitterButton + copyButton;
	//Related to copyButton button
	document.getElementById ("copy-share-link").addEventListener ("mouseover", () => { document.body.style.cursor = "pointer"});
	document.getElementById ("copy-share-link").addEventListener ("mouseout", () => { document.body.style.cursor = "default"});
	document.getElementById ("copy-share-link").addEventListener ("click", () => {
	    var tmpInput = document.createElement ("input");
	    let urlText = window.location.href;
	    document.body.appendChild (tmpInput);
	    tmpInput.value = urlText;
	    tmpInput.select ();
	    document.execCommand ("copy");
	    document.body.removeChild (tmpInput);
	    alert ("The link for this article has been copied to your clipboard!");
	});
}

js/write.js
document.addEventListener ('DOMContentLoaded', () => {
    //Start ContentTools
    startEditor ();

    //To filter unwanted characters from elements content
    function filterCharacters (str) {
        return (str.replace(/<(?:.|\n)*?>/gm, '').trim ());
    }

    //Load database file
    function loadDatabase (file, callback) {
        let xhr = new XMLHttpRequest ();
        xhr.overrideMimeType ("application/json");
        xhr.open ("GET", file, true);
        xhr.onreadystatechange = () => {
            if (xhr.readyState === XMLHttpRequest.DONE) {
                if (xhr.status === 0 || (xhr.status >= 200 && xhr.status < 400)) {
                    try {
                        callback (JSON.parse (xhr.responseText));
                    } catch (e) {
                        callback (null);
                    }
                } else {
                    callback (null);
                }
            }
        }
        xhr.send (null);
    }

    //Download content as file
    function downloadFile (filename, dataType, content) {
        console.log ('hi');
        var element = document.createElement ("a");
        element.setAttribute ("href", "data:" + dataType +";charset=UTF-8," + encodeURIComponent(content));
        element.setAttribute ("download", filename);
        element.style.display = "none";
        document.body.appendChild (element);
        element.click ();
        document.body.removeChild (element);
    }

    //Set article date
    let today = new Date ();
    let monthNames = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
    let day = today.getDate ();
    let monthIndex = today.getMonth ();
    let year = today.getFullYear ();
    let finalDate = day + ' ' + monthNames[monthIndex] + ' ' + year;
    let dateParagraph = document.createElement ('p');
    dateParagraph.innerHTML = finalDate;
    document.getElementById ('article-date').innerHTML = '';
    document.getElementById ('article-date').appendChild (dateParagraph);

    //Add event listeners for state changes in article banner url, doing so will automatically update the image while editing the url
    let image_url_input = document.getElementById ('article-image-url');
    let article_image = document.getElementById ('article-image');
    ['change', 'keydown', 'keypress', 'keyup', 'mousedown', 'click', 'mouseup'].forEach ((event) => {
        image_url_input.addEventListener (event, () => {
            let url = image_url_input.value;
            if (url) {
                article_image.style.backgroundImage = 'url('+ url + ')';
                article_image.style.display = 'inline-block';
            } else {
                article_image.style.display = 'none';
            }
        })
    });

    //Get new article data and save to database
    document.getElementById ('finish-editing').addEventListener ('click', () => {
        //Get data to go to database from key components on the page
        let article_title = filterCharacters (document.getElementById ('article-title').firstElementChild.innerHTML);
        let article_subtitle = filterCharacters (document.getElementById ('article-subtitle').firstElementChild.innerHTML);
        let article_date = filterCharacters (document.getElementById ('article-date').firstElementChild.innerHTML);

        let article_tags = new Array ();
        let tags = document.getElementById ('article-tags').firstElementChild.childNodes;
        Array.prototype.forEach.call (tags, (tag) => {
            let clean_tag = tag.innerHTML + '';
            clean_tag = clean_tag.replace (/<(?:.|\n)*?>/gm, '').trim ();
            clean_tag = clean_tag.replace (/\s\s+/g, '');
            if (clean_tag) article_tags.push (clean_tag);
        });
        
        let article_image = filterCharacters (document.getElementById ('article-image').style.backgroundImage.slice (5, -2));
        let article_author = filterCharacters (document.getElementById ('article-author').firstElementChild.innerHTML);
        let article_summary = btoa (unescape (encodeURIComponent (document.getElementById ('article-text').firstElementChild.innerHTML)));
        let article_text = btoa (unescape (encodeURIComponent (document.getElementById ('article-text').innerHTML)));

        loadDatabase ('../data/articles.json', (articles) => {
            //The id for the new article
            let new_id = 0;
            //Check if articles equals null which means no database
            if (articles && articles[0]) {
                //Verify database to use unique id for the new article
                new_id = parseInt (articles[0].articleID) +1;
                let id_conflict = false;
                do {
                    articles.forEach ((article) => {
                        if (new_id == article.articleID) {
                            id_conflict = true;
                            new_id++;
                        }
                    });
                } while (id_conflict == true);
            }

            //Join new article components
            let new_article = {
                "articleID": new_id,
                "articleTitle": article_title,
                "articleSubtitle": article_subtitle,
                "articleDate": article_date,
                "articleTags": article_tags,
                "articleImage": article_image,
                "articleAuthor": article_author,
                "articleSummary": article_summary,
                "articleText": article_text
            };
            //If articles is not null adds new_article to database
            //Else creates a new database file
            if (articles) articles.unshift (new_article);
            else articles = [new_article];
            //Json conversion and download
            articles = JSON.stringify (articles, null, 4);
            downloadFile ('articles.json', 'application/json', articles);
        });
    });
});